File: azkar/kotlin/shared/src/commonMain/kotlin/com/wildduck/azkar/ui/SchedulePage.kt
// The interval, and the morning, evening and quiet-hours times.
package com.wildduck.azkar.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import com.wildduck.azkar.app.Features
import com.wildduck.azkar.app.UiState
import com.wildduck.azkar.app.WindowKey
import com.wildduck.azkar.core.Config
import com.wildduck.azkar.core.TimeWindow
import kotlin.math.floor
import kotlin.math.roundToInt

@Composable
fun SchedulePage(ui: UiState, features: Features, actions: AzkarActions) {
    val config = ui.config
    Page("Schedule") {
        Section(footer = "A new card appears on this interval while there's room on screen.") {
            StepperRow(
                title = "Remind me every",
                color = Accents.teal,
                value = "${trim(config.intervalMinutes)} min",
                range = 1..120,
                current = config.intervalMinutes.roundToInt(),
                onChange = { actions.setConfig(config.copy(intervalMinutes = it.toDouble())) },
            )
        }

        WindowSection(
            key = WindowKey.Sabah,
            color = Accents.orange,
            footer = "Goes through the ${ui.library.sabah.size} morning azkar in order, once a day.",
            config = config,
            actions = actions,
        )
        WindowSection(
            key = WindowKey.Masaa,
            color = Accents.purple,
            footer = "Goes through the ${ui.library.masaa.size} evening azkar in order, once a day.",
            config = config,
            actions = actions,
        )
        WindowSection(
            key = WindowKey.Quiet,
            color = Accents.indigo,
            footer = "No new cards during these hours.",
            config = config,
            actions = actions,
        )

        Footnote(
            "Between and after the morning and evening times — and once their list is done — you get general " +
                "azkar" + (if (config.repeatGeneral) "" else ", one tap each") + ".",
        )
    }
}

/** One daily window: a switch, and its two times while it is on. */
@Composable
private fun WindowSection(
    key: WindowKey,
    color: Color,
    footer: String,
    config: Config,
    actions: AzkarActions,
) {
    val window = key.of(config)
    Section(footer = footer) {
        SwitchRow(
            title = key.label,
            color = color,
            checked = window != null,
            onChange = { on ->
                actions.setConfig(key.set(config, if (on) actions.rememberedWindow(key) else null))
            },
        )
        if (window != null) {
            Row(
                Modifier.fillMaxWidth().padding(start = 48.dp, end = 14.dp, bottom = 12.dp),
                horizontalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                ClockField(
                    label = "From",
                    minutes = window.start,
                    onChange = { actions.setConfig(key.set(config, TimeWindow(it, window.end))) },
                    modifier = Modifier.testTag("clock:${key.name}:from"),
                )
                ClockField(
                    label = "To",
                    minutes = window.end,
                    onChange = { actions.setConfig(key.set(config, TimeWindow(window.start, it))) },
                    modifier = Modifier.testTag("clock:${key.name}:to"),
                )
                Text(
                    if (window.end <= window.start) "Runs past midnight" else "",
                    Modifier.padding(top = 18.dp),
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    style = MaterialTheme.typography.bodySmall,
                )
            }
        }
    }
}

/** 3.0 -> "3", 2.5 -> "2.5". */
private fun trim(minutes: Double): String =
    if (minutes == floor(minutes)) minutes.toInt().toString() else minutes.toString()

File: azkar/apple/macOS/Window/SchedulePage.swift
import SwiftUI

/// The interval, and the morning, evening and quiet-hours times.
struct SchedulePage: View {
  @ObservedObject var model: SettingsModel

  var body: some View {
    Form {
      Section {
        Stepper(value: $model.config.intervalMinutes, in: 1...120, step: 1) {
          Row(title: "Remind me every", symbol: "timer", color: .teal) {
            Text("\(model.config.intervalMinutes.formatted()) min").monospacedDigit().foregroundStyle(.secondary)
          }
        }
      } footer: {
        Footnote("A new card appears on this interval while there's room on screen.")
      }

      window(
        \.sabah, "Morning azkar", "sunrise.fill", .orange,
        "Goes through the \(model.library.sabah.count) morning azkar in order, once a day.")
      window(
        \.masaa, "Evening azkar", "moon.stars.fill", .purple,
        "Goes through the \(model.library.masaa.count) evening azkar in order, once a day.")
      window(\.quietHours, "Quiet hours", "moon.zzz.fill", .indigo, "No new cards during these hours.")

      Section {
        Label(
          "Between and after the morning and evening times — and once their list is done — you get general azkar\(model.config.repeatGeneral ? "" : ", one tap each").",
          systemImage: "info.circle"
        )
        .foregroundStyle(.secondary)
      }
    }
    .formStyle(.grouped)
    .navigationTitle("Schedule")
  }

  private func window(
    _ w: WritableKeyPath<Config, TimeWindow?>, _ title: String, _ symbol: String,
    _ color: Color, _ footer: String
  ) -> some View {
    Section {
      Toggle(isOn: Binding(get: { model.isEnabled(w) }, set: { model.setEnabled(w, $0) })) {
        Row(title: title, symbol: symbol, color: color) { EmptyView() }
      }
      .toggleStyle(.switch)
      if model.isEnabled(w) {
        DatePicker(
          "From",
          selection: Binding(get: { model.time(w, start: true) }, set: { model.setTime(w, start: true, to: $0) }),
          displayedComponents: .hourAndMinute)
        DatePicker(
          "To",
          selection: Binding(get: { model.time(w, start: false) }, set: { model.setTime(w, start: false, to: $0) }),
          displayedComponents: .hourAndMinute)
      }
    } footer: {
      Footnote(footer)
    }
  }
}
